synapse/crypto/context_factory.py

```diff
@@ -48,3 +51,45 @@ def configure_context(context, config):
 
     def getContext(self):
         return self._context
+
+
+@implementer(IOpenSSLClientConnectionCreator)
+class ClientTLSOptions(object):
+    """
+    Client creator for TLS without certificate identity verification. This is a
+    copy of twisted.internet._sslverify.ClientTLSOptions with the identity
+    verification left out. For documentation, see the twisted documentation.
+    """
+
+    def __init__(self, hostname, ctx):
+        self._ctx = ctx
+        self._hostname = hostname
+        self._hostnameBytes = _idnaBytes(hostname)
+        ctx.set_info_callback(
+            _tolerateErrors(self._identityVerifyingInfoCallback)
+        )
+
+    def clientConnectionForTLS(self, tlsProtocol):
+        context = self._ctx
+        connection = SSL.Connection(context, None)
+        connection.set_app_data(tlsProtocol)
+        return connection
+
+    def _identityVerifyingInfoCallback(self, connection, where, ret):
+        if where & SSL.SSL_CB_HANDSHAKE_START:
+            connection.set_tlsext_host_name(self._hostnameBytes)
+
+
+class ClientTLSOptionsFactory(object):
+    """Factory for Twisted ClientTLSOptions that are used to make connections
+    to remote servers for federation."""
+
+    def __init__(self, config):
+        # We don't use config options yet
+        pass
+
+    def get_options(self, host):
+        return ClientTLSOptions(
+            unicode(host),
```

Comment:
we're in the process of porting to python 3, and this won't work. Maybe you want `host.encode('ascii')` ?

Reply:
ClientTLSOptions should get a unicode string, so the python 3 equivalent of `unicode(host)` would be `str(host)`. If the same code should run in python 2 and 3, assigning `str` to `unicode` for python 3 is a straightforward solution. Is this OK for you?

```
if sys.version_info.major >= 3:
    unicode = str
```

Reply:
sorry, I was being stupid here. Let's not go defining `unicode` though.

I think you actually want `host.decode('ascii')` ?

Reply:
Is there a specific reason you want to use `decode('ascii')`, otherwise I would suggest `decode('utf-8')` to keep it more generic? 

Just quickly looking at `synapse.http.endpoint.matrix_federation_endpoint` (where `ClientTLSOptionsFactory` is actually used), it seems to me that this function will have more problems with bytes vs str when running in python 3 than just this one. Probably to be looked at in another PR, as this would require some more investigation.

Reply:
> Is there a specific reason you want to use decode('ascii'), otherwise I would suggest decode('utf-8') to keep it more generic?

Well, I think we're expecting a plain ascii hostname here. If it's *not* plain ascii, something is going wrong.

And yes, there are currently other python 3 problems in this area. @hawkowl  is working on python 3 support...

Reply:
I see you already merged this. I did the commit with utf-8, so if you prefer ascii it might be best to change it then.